crates/sui-types/src/object.rs

```diff
@@ -398,6 +400,68 @@ impl MoveObject {
     }
 }
 
+// Helpers for extracting Coin<T> balances for all T
+impl MoveObject {
+    fn is_balance(s: &StructTag) -> Option<&TypeTag> {
+        (Balance::is_balance(s) && s.type_params.len() == 1).then_some(&s.type_params[0])
+    }
+
+    /// Get the total balances for all Coin<T> embedded in `self`.
+    pub fn get_coin_balances(
+        &self,
+        layout_resolver: &mut dyn LayoutResolver,
+    ) -> Result<HashMap<TypeTag, u64>, SuiError> {
+        let mut balances = HashMap::default();
```

Comment:
We might want to consider collection type with deterministic iteration order since it's easy to (e.g.) imagine indexer or RPC-populating logic iterating over this. But I think we can also reasonably make that the caller's problem.

Reply:
Easy enough to just use a BTreeMap